#Gensim library

In [ ]:
!pip install gensim
!pip install python-Levenshtein

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 162.6/162.6 kB 6.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 40.7 MB/s eta 0:00:00


In [ ]:
import gensim
import pandas as pd

In [ ]:
df = pd.read_json("/content/Cell_Phones_and_Accessories_5.json", lines=True)
df

reviewerID        asin       reviewerName helpful  \
0       A30TL5EWN6DFXT  120401325X          christina  [0, 0]   
1        ASY55RVNIL0UD  120401325X           emily l.  [0, 0]   
2       A2TMXE2AFO7ONB  120401325X              Erica  [0, 0]   
3        AWJ0WZQYMYFQ4  120401325X                 JM  [4, 4]   
4        ATX7CZYFXI1KW  120401325X   patrice m rogoza  [2, 3]   
...                ...         ...                ...     ...   
194434  A1YMNTFLNDYQ1F  B00LORXVUE    eyeused2loveher  [0, 0]   
194435  A15TX8B2L8B20S  B00LORXVUE       Jon Davidson  [0, 0]   
194436  A3JI7QRZO1QG8X  B00LORXVUE  Joyce M. Davidson  [0, 0]   
194437  A1NHB2VC68YQNM  B00LORXVUE     Nurse Farrugia  [0, 0]   
194438  A1AG6U022WHXBF  B00LORXVUE     Trisha Crocker  [0, 0]   

                                               reviewText  overall  \
0       They look good and stick good! I just don't li...        4   
1       These stickers work like the review says they ...        5   
2       These are awesome and make my phone look so st...        5   
3       Item arrived in great time and was in perfect ...        4   
4       awesome! stays on, and looks great. can be use...        5   
...                                                   ...      ...   
194434  Works great just like my original one. I reall...        5   
194435  Great product. Great packaging. High quality a...        5   
194436  This is a great cable, just as good as the mor...        5   
194437  I really like it becasue it works well with my...        5   
194438  product as described, I have wasted a lot of m...        5   

                                                  summary  unixReviewTime  \
0                                              Looks Good      1400630400   
1                                   Really great product.      1389657600   
2                                          LOVE LOVE LOVE      1403740800   
3                                                   Cute!      1382313600   
4               leopard home button sticker for iphone 4s      1359849600   
...                                                   ...             ...   
194434                           This works just perfect!      1405900800   
194435           Great replacement cable. Apple certified      1405900800   
194436                                       Real quality      1405900800   
194437  I really like it becasue it works well with my...      1405814400   
194438              I have wasted a lot of money on cords      1405900800   

         reviewTime  
0       05 21, 2014  
1       01 14, 2014  
2       06 26, 2014  
3       10 21, 2013  
4        02 3, 2013  
...             ...  
194434  07 21, 2014  
194435  07 21, 2014  
194436  07 21, 2014  
194437  07 20, 2014  
194438  07 21, 2014  

[194439 rows x 9 columns]

In [ ]:
df.shape

(194439, 9)

In [ ]:
review_text = df.reviewText.apply(gensim.utils.simple_preprocess)

In [ ]:
review_text

0         [they, look, good, and, stick, good, just, don...
1         [these, stickers, work, like, the, review, say...
2         [these, are, awesome, and, make, my, phone, lo...
3         [item, arrived, in, great, time, and, was, in,...
4         [awesome, stays, on, and, looks, great, can, b...
                                ...                        
194434    [works, great, just, like, my, original, one, ...
194435    [great, product, great, packaging, high, quali...
194436    [this, is, great, cable, just, as, good, as, t...
194437    [really, like, it, becasue, it, works, well, w...
194438    [product, as, described, have, wasted, lot, of...
Name: reviewText, Length: 194439, dtype: object

In [ ]:
review_text.loc[0]

['they',
 'look',
 'good',
 'and',
 'stick',
 'good',
 'just',
 'don',
 'like',
 'the',
 'rounded',
 'shape',
 'because',
 'was',
 'always',
 'bumping',
 'it',
 'and',
 'siri',
 'kept',
 'popping',
 'up',
 'and',
 'it',
 'was',
 'irritating',
 'just',
 'won',
 'buy',
 'product',
 'like',
 'this',
 'again']

In [ ]:
df.reviewText.loc[0]

"They look good and stick good! I just don't like the rounded shape because I was always bumping it and Siri kept popping up and it was irritating. I just won't buy a product like this again"

In [ ]:
model = gensim.models.Word2Vec(
    window=10,
    min_count=2,
    workers=4,
)

In [ ]:
model.build_vocab(review_text, progress_per=1000)

In [ ]:
model.train(review_text, total_examples=model.corpus_count, epochs=model.epochs)

(61503089, 83868975)

In [ ]:
model.save("./word2vec-amazon-cell-accessories-reviews-short.model")

In [ ]:
model.wv.most_similar("bad")

[('terrible', 0.6792085766792297),
 ('horrible', 0.6324943900108337),
 ('shabby', 0.6171278953552246),
 ('good', 0.594677746295929),
 ('crappy', 0.5586413145065308),
 ('poor', 0.5427101850509644),
 ('awful', 0.5299966931343079),
 ('cheap', 0.5275050401687622),
 ('funny', 0.5212780833244324),
 ('sad', 0.513801634311676)]

In [ ]:
model.wv.similarity(w1="cheap", w2="inexpensive")

0.5556282

In [ ]:
model.wv.similarity(w1="great", w2="good")

0.7843319

#Small Image Classification Using Simple Aritifical Neural Network: GPU Benchmarking

In [ ]:
import os
os.environ["CUDA_VISIBLE_DEVICES"]="4"

In [ ]:
import tensorflow as tf
from tensorflow import keras
import numpy as np

In [ ]:
tf.config.experimental.list_physical_devices()

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]

In [ ]:
tf.__version__

'2.17.0'

In [ ]:
tf.test.is_built_with_cuda()

True

In [ ]:
(X_train, y_train), (X_test,y_test) = tf.keras.datasets.cifar10.load_data()

In [ ]:
X_train.shape

(50000, 32, 32, 3)

In [ ]:
y_train.shape

(50000, 1)

In [ ]:
classes = ["airplane","automobile","bird","cat","deer","dog","frog","horse","ship","truck"]

In [ ]:
classes[y_train[3][0]]

'deer'

In [ ]:
y_train[:3]

array([[6],
       [9],
       [9]], dtype=uint8)

In [ ]:
y_test.shape

(10000, 1)

In [ ]:
X_train.shape

(50000, 32, 32, 3)

In [ ]:
X_train_scaled = X_train / 255
X_test_scaled = X_test / 255

In [ ]:
y_train_categorical = keras.utils.to_categorical(
    y_train, num_classes=10
)
y_test_categorical = keras.utils.to_categorical(
    y_test, num_classes=10
)

In [ ]:
y_train[0:5]

array([[6],
       [9],
       [9],
       [4],
       [1]], dtype=uint8)

In [ ]:
y_train_categorical[0:5]

array([[0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [ ]:
def get_model():
    model = keras.Sequential([
            keras.layers.Flatten(input_shape=(32,32,3)),
            keras.layers.Dense(3000, activation='relu'),
            keras.layers.Dense(1000, activation='relu'),
            keras.layers.Dense(10, activation='sigmoid')
        ])

    model.compile(optimizer='SGD',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
    return model

In [ ]:
train_tf_dataset = tf.data.Dataset.from_tensor_slices((X_train_scaled, y_train_categorical))
test_tf_dataset = tf.data.Dataset.from_tensor_slices((X_test_scaled, y_test_categorical))

In [ ]:
strategy = tf.distribute.MirroredStrategy()

In [ ]:
strategy.num_replicas_in_sync

1

In [ ]:
BATCH_SIZE_PER_REPLICA = 1000
BATCH_SIZE = BATCH_SIZE_PER_REPLICA * strategy.num_replicas_in_sync


train_dataset = train_tf_dataset.batch(BATCH_SIZE).prefetch(tf.data.AUTOTUNE)
test_dataset = test_tf_dataset.batch(BATCH_SIZE)

In [ ]:
%%timeit -n1 -r1
with strategy.scope():
    gpu_model = get_model()
    gpu_model.fit(train_dataset, epochs=5)

Epoch 1/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 52s 1000ms/step - accuracy: 0.2080 - loss: 2.2074
Epoch 2/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 49s 965ms/step - accuracy: 0.3156 - loss: 1.9847
Epoch 3/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 82s 974ms/step - accuracy: 0.3438 - loss: 1.9059
Epoch 4/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 82s 981ms/step - accuracy: 0.3603 - loss: 1.8584
Epoch 5/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 51s 1s/step - accuracy: 0.3694 - loss: 1.8257
5min 49s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


In [ ]:
%%timeit -n1 -r1
with tf.device('/CPU:0'):
    cpu_model = get_model()
    cpu_model.fit(train_dataset, epochs=5)

Epoch 1/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 58s 1s/step - accuracy: 0.1980 - loss: 2.2061
Epoch 2/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 49s 987ms/step - accuracy: 0.3068 - loss: 1.9892
Epoch 3/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 49s 978ms/step - accuracy: 0.3358 - loss: 1.9113
Epoch 4/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 82s 986ms/step - accuracy: 0.3527 - loss: 1.8647
Epoch 5/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 51s 1s/step - accuracy: 0.3655 - loss: 1.8319
4min 50s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


#Tensorflow Input PipeLine

In [ ]:
import tensorflow as tf

In [ ]:
daily_sales_numbers = [21, 22, -108, 31, -1, 32, 34,31]

tf_dataset = tf.data.Dataset.from_tensor_slices(daily_sales_numbers)
tf_dataset

<_TensorSliceDataset element_spec=TensorSpec(shape=(), dtype=tf.int32, name=None)>

In [ ]:
for sales in tf_dataset:
    print(sales.numpy())

21
22
-108
31
-1
32
34
31


In [ ]:
for sales in tf_dataset.as_numpy_iterator():
    print(sales)

21
22
-108
31
-1
32
34
31


In [ ]:
for sales in tf_dataset.take(3):
    print(sales.numpy())

21
22
-108


In [ ]:
tf_dataset = tf_dataset.filter(lambda x: x>0)
for sales in tf_dataset.as_numpy_iterator():
    print(sales)

21
22
31
32
34
31


In [ ]:
tf_dataset = tf_dataset.map(lambda x: x*72)
for sales in tf_dataset.as_numpy_iterator():
    print(sales)

1512
1584
2232
2304
2448
2232


In [ ]:
tf_dataset = tf_dataset.shuffle(2)
for sales in tf_dataset.as_numpy_iterator():
    print(sales)

1584
1512
2304
2448
2232
2232


In [ ]:
for sales_batch in tf_dataset.batch(2):
    print(sales_batch.numpy())

[1584 2232]
[2304 2448]
[1512 2232]


In [ ]:
tf_dataset = tf.data.Dataset.from_tensor_slices(daily_sales_numbers)

tf_dataset = tf_dataset.filter(lambda x: x>0).map(lambda y: y*72).shuffle(2).batch(2)
for sales in tf_dataset.as_numpy_iterator():
    print(sales)

[1584 1512]
[2304 2232]
[2232 2448]


In [ ]:
!ls images

ls: cannot access 'images': No such file or directory


In [ ]:
import os
print(os.getcwd())

/content


In [ ]:
images_ds = tf.data.Dataset.list_files('/content*', shuffle=False)

In [ ]:
image_count = len(images_ds)
image_count

1

In [ ]:
type(images_ds)

tensorflow.python.data.ops.from_tensor_slices_op._TensorSliceDataset

In [ ]:
for file in images_ds.take(3):
    print(file.numpy())

b'/content'


In [ ]:
images_ds = images_ds.shuffle(200)
for file in images_ds.take(3):
    print(file.numpy())

b'/content'


In [ ]:
class_names = ["cat","dog"]

In [ ]:
train_size = int(image_count*0.8)
train_ds = images_ds.take(train_size)
test_ds = images_ds.skip(train_size)

In [ ]:
len(train_ds)

0

In [ ]:
len(test_ds)

1

In [ ]:
def get_label(file_path):
    import os
    parts = tf.strings.split(file_path, os.path.sep)
    return parts[-2]

In [ ]:
get_label("images\\dog\\20 Reasons Why Cats Make the Best Pets....jpg")

InvalidArgumentError: {{function_node __wrapped__StridedSlice_device_/job:localhost/replica:0/task:0/device:CPU:0}} slice index -1 of dimension 0 out of bounds. [Op:StridedSlice] name: strided_slice/

In [ ]:
def process_image(file_path):
    label = get_label(file_path)
    img = tf.io.read_file(file_path) # load the raw data from the file as a string
    img = tf.image.decode_jpeg(img)
    img = tf.image.resize(img, [128, 128])
    return img, label

In [ ]:
img, label = process_image("images\\cat\\20 Reasons Why Cats Make the Best Pets....jpg")
img.numpy()[:2]

In [ ]:
train_ds = train_ds.map(process_image)
test_ds = test_ds.map(process_image)

In [ ]:
for image, label in train_ds.take(1):
    print("****",image)
    print("****",label)

In [ ]:
def scale(image, label):
    return image/255, label

In [ ]:
train_ds = train_ds.map(scale)

In [ ]:
for image, label in train_ds.take(5):
    print("****Image: ",image.numpy()[0][0])
    print("****Label: ",label.numpy())

#BERT MODELS

In [ ]:
!pip install tensorflow-text

In [ ]:
import tensorflow_hub as hub
import tensorflow_text as text

In [ ]:
preprocess_url = "https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3"
encoder_url = "https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4"

In [ ]:
bert_preprocess_model = hub.KerasLayer(preprocess_url)
text_test = ['nice movie indeed','I love python programming']

In [ ]:
text_preprocessed = bert_preprocess_model(text_test)
text_preprocessed.keys()

dict_keys(['input_mask', 'input_word_ids', 'input_type_ids'])

In [ ]:
text_preprocessed['input_mask']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      dtype=int32)>

In [ ]:
text_preprocessed['input_type_ids']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      dtype=int32)>

In [ ]:
text_preprocessed['input_word_ids']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[  101,  3835,  3185,  5262,   102,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0, 

101 --> CLS token

102 --> SEP token

BERT uses CLS as a special token at the begining of each setence whereas SEP as a special token to separate two sentences or end sinle sentence

In [ ]:
bert_model = hub.KerasLayer(encoder_url)

In [ ]:
bert_results = bert_model(text_preprocessed)

In [ ]:
bert_results.keys()

dict_keys(['sequence_output', 'pooled_output', 'encoder_outputs', 'default'])

In [ ]:
bert_results['sequence_output']

<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
array([[[ 0.07292047,  0.08567817,  0.14476831, ..., -0.09677088,
          0.08722127,  0.07711085],
        [ 0.1783942 , -0.19006078,  0.5034952 , ..., -0.05869824,
          0.32717073, -0.15578493],
        [ 0.18701449, -0.43388748, -0.4887515 , ..., -0.15502778,
          0.00145182, -0.24471016],
        ...,
        [ 0.12083091,  0.12884252,  0.46453542, ...,  0.0737552 ,
          0.17441946,  0.16522075],
        [ 0.07967882, -0.01190652,  0.502254  , ...,  0.13777725,
          0.21002197,  0.00624588],
        [-0.07212701, -0.28303462,  0.59033376, ...,  0.475519  ,
          0.16668482, -0.08920349]],

       [[-0.0790059 ,  0.36335096, -0.21101595, ..., -0.17183761,
          0.16299714,  0.6724268 ],
        [ 0.27883548,  0.43716297, -0.35764822, ..., -0.04463654,
          0.38315183,  0.5887981 ],
        [ 1.2037671 ,  1.0727016 ,  0.48408663, ...,  0.24921048,
          0.407309  ,  0.40481755],
        ...,

In [ ]:
bert_results['pooled_output']

<tf.Tensor: shape=(2, 768), dtype=float32, numpy=
array([[-0.79177403, -0.21411912,  0.49769545, ...,  0.24465235,
        -0.47334474,  0.81758696],
       [-0.91712314, -0.47935176, -0.7865697 , ..., -0.6175175 ,
        -0.71026856,  0.921843  ]], dtype=float32)>

In [ ]:
len(bert_results['encoder_outputs'])

12

Since we are using BERT base model it has 12 encoder layers, that's why above the length of encoder_output is 12. Read more about purpose of each element here: https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4

In [ ]:
bert_results['encoder_outputs'][-1] == bert_results['sequence_output']

<tf.Tensor: shape=(2, 128, 768), dtype=bool, numpy=
array([[[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]],

       [[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]]])>

In [ ]:
bert_results['encoder_outputs']

[<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
 array([[[ 0.1290144 ,  0.00644745, -0.03614967, ...,  0.04999624,
           0.06149195, -0.02657542],
         [ 1.1753386 ,  1.2140785 ,  1.1569982 , ...,  0.11634357,
          -0.3585537 , -0.40490186],
         [ 0.03859041,  0.5386997 , -0.2108978 , ...,  0.21858197,
           0.7260168 , -1.1158606 ],
         ...,
         [-0.07587015, -0.25421894,  0.7075512 , ...,  0.50541997,
          -0.1887868 ,  0.15028352],
         [-0.16066599, -0.28089687,  0.5759707 , ...,  0.5275855 ,
          -0.11141382,  0.02887543],
         [-0.04428157, -0.20279568,  0.59093547, ...,  0.8133835 ,
          -0.39075816, -0.02601742]],
 
        [[ 0.18903585,  0.02752551, -0.06513745, ..., -0.00620203,
           0.15053892,  0.0316545 ],
         [ 0.5916147 ,  0.7589138 , -0.07240658, ...,  0.61903954,
           0.8292891 ,  0.16161972],
         [ 1.4460826 ,  0.44602674,  0.40990272, ...,  0.48255897,
           0.6269117 ,  0.13

#BERT tutorial: Classify spam vs no spam emails

In [2]:
!pip install tensorflow-text

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.2/5.2 MB 42.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 615.3/615.3 MB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 51.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 90.7 MB/s eta 0:00:00
  Attempting uninstall: tensorboard
    Found existing installation: tensorboard 2.17.0
    Uninstalling tensorboard-2.17.0:
      Successfully uninstalled tensorboard-2.17.0
  Attempting uninstall: keras
    Found existing installation: keras 3.4.1
    Uninstalling keras-3.4.1:
      Successfully uninstalled keras-3.4.1
  Attempting uninstall: tensorflow
    Found existing installation: tensorflow 2.17.0
    Uninstalling tensorflow-2.17.0:
      Successfully uninstalled tensorflow-2.17.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tf-keras 2.17.0 requ

In [1]:
import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_text as text

In [2]:
import pandas as pd

df = pd.read_csv("/content/spam.csv")
df.head(5)

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [3]:
df.groupby('Category').describe()

Message                                                            \
           count unique                                                top   
Category                                                                     
ham         4825   4516                             Sorry, I'll call later   
spam         747    641  Please call our customer service representativ...   

               
         freq  
Category       
ham        30  
spam        4

In [4]:
df['Category'].value_counts()

Category
ham     4825
spam     747
Name: count, dtype: int64

In [5]:
df_spam = df[df['Category']=='spam']
df_spam.shape

(747, 2)

In [6]:
df_ham = df[df['Category']=='ham']
df_ham.shape

(4825, 2)

In [7]:
df_ham_downsampled = df_ham.sample(df_spam.shape[0])
df_ham_downsampled.shape

(747, 2)

In [8]:
df_balanced = pd.concat([df_ham_downsampled, df_spam])
df_balanced.shape

(1494, 2)

In [9]:
df_balanced['Category'].value_counts()

Category
ham     747
spam    747
Name: count, dtype: int64

In [10]:
df_balanced['spam']=df_balanced['Category'].apply(lambda x: 1 if x=='spam' else 0)
df_balanced.sample(5)

Category                                            Message  spam
4635      ham        These won't do. Have to move on to morphine     0
2916      ham  Nothing lor... A bit bored too... Then y dun u...     0
2525     spam  FREE entry into our £250 weekly comp just send...     1
4509      ham                           Not able to do anything.     0
1072     spam  URGENT! We are trying to contact U. Todays dra...     1

In [11]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df_balanced['Message'],df_balanced['spam'], stratify=df_balanced['spam'])

In [12]:
X_train.head(4)

1594    PRIVATE! Your 2003 Account Statement for shows...
605                                Meet after lunch la...
4796    Saw Guys and Dolls last night with Patrick Swa...
1963    it to 80488. Your 500 free text messages are v...
Name: Message, dtype: object

In [13]:
bert_preprocess = hub.KerasLayer("https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3")
bert_encoder = hub.KerasLayer("https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4")

In [14]:
def get_sentence_embeding(sentences):
    preprocessed_text = bert_preprocess(sentences)
    return bert_encoder(preprocessed_text)['pooled_output']

get_sentence_embeding([
    "500$ discount. hurry up",
    "Bhavin, are you up for a volleybal game tomorrow?"]
)

<tf.Tensor: shape=(2, 768), dtype=float32, numpy=
array([[-0.8435168 , -0.5132724 , -0.8884571 , ..., -0.7474884 ,
        -0.7531473 ,  0.91964483],
       [-0.87208354, -0.50543964, -0.94446665, ..., -0.85847497,
        -0.7174535 ,  0.88082975]], dtype=float32)>

In [15]:
e = get_sentence_embeding([
    "banana",
    "grapes",
    "mango",
    "jeff bezos",
    "elon musk",
    "bill gates"
]
)

In [16]:
from sklearn.metrics.pairwise import cosine_similarity
cosine_similarity([e[0]],[e[1]])

array([[0.9911088]], dtype=float32)

In [17]:
cosine_similarity([e[0]],[e[3]])

array([[0.84703827]], dtype=float32)

In [18]:
cosine_similarity([e[3]],[e[4]])

array([[0.9872036]], dtype=float32)

In [23]:
import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_text as text

# Load BERT preprocessing and encoder layers
bert_preprocess = hub.KerasLayer("https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3", name="bert_preprocess")
bert_encoder = hub.KerasLayer("https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/3", trainable=True, name="bert_encoder")

# Define the input layer
text_input = tf.keras.layers.Input(shape=(), dtype=tf.string, name='text')
@tf.function
def encode_text(text):
    # Preprocess and encode the input text
    preprocessed_text = bert_preprocess(text)
    return bert_encoder(preprocessed_text)['pooled_output']

# Specify the output shape in the Lambda layer
bert_output = tf.keras.layers.Lambda(encode_text, output_shape=(768,))(text_input)

In [28]:
# Add classification layers
dropout_layer = tf.keras.layers.Dropout(0.1, name="dropout")(bert_output)
output_layer = tf.keras.layers.Dense(1, activation='sigmoid', name="output")(dropout_layer)

# Create the final model
model = tf.keras.Model(inputs=[text_input], outputs=[output_layer])

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [29]:
# Display the model summary
model.summary()

Model: "functional_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ text (InputLayer)                    │ (None)                      │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ lambda_1 (Lambda)                    │ (None, 768)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout (Dropout)                    │ (None, 768)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ output (Dense)                       │ (None, 1)                   │             769 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 769 (3.00 KB)

 Trainable params: 769 (3.00 KB)

 Non-trainable params: 0 (0.00 B)

In [30]:
len(X_train)

1120

In [31]:
METRICS = [
      tf.keras.metrics.BinaryAccuracy(name='accuracy'),
      tf.keras.metrics.Precision(name='precision'),
      tf.keras.metrics.Recall(name='recall')
]

model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=METRICS)

In [33]:
model.fit(X_train, y_train, epochs=1)

35/35 ━━━━━━━━━━━━━━━━━━━━ 545s 15s/step - accuracy: 0.6489 - loss: 0.6481 - precision: 0.6681 - recall: 0.5740


In [34]:
model.evaluate(X_test, y_test)

12/12 ━━━━━━━━━━━━━━━━━━━━ 180s 14s/step - accuracy: 0.7968 - loss: 0.5366 - precision: 0.7284 - recall: 0.9560


[0.537118136882782, 0.8101603984832764, 0.7396694421768188, 0.9572192430496216]

In [35]:
y_predicted = model.predict(X_test)
y_predicted = y_predicted.flatten()

12/12 ━━━━━━━━━━━━━━━━━━━━ 178s 15s/step


In [36]:
y_predicted = model.predict(X_test)
y_predicted = y_predicted.flatten()

12/12 ━━━━━━━━━━━━━━━━━━━━ 176s 15s/step


In [ ]:
from matplotlib import pyplot as plt
import seaborn as sn
sn.heatmap(cm, annot=True, fmt='d')
plt.xlabel('Predicted')
plt.ylabel('Truth')

In [ ]:
print(classification_report(y_test, y_predicted))